File: backend/app/services/customs_release_service.py
from datetime import datetime

from app.extensions import db
from app.models.package import Package
from app.services.billing_calculations import publish_ready_for_pickup_bill
from app.services.billing_service import request_package_invoice
from app.services.package_service import add_package_event


def release_package_from_customs(
    package: Package,
    *,
    estimated_freight_jmd: float | None = None,
    duties_jmd: float | None = None,
    handling_jmd: float | None = None,
    other_fees_jmd: float | None = None,
    note: str | None = None,
) -> Package:
    if package.status != "customs":
        raise ValueError(f"{package.tracking_number} is not in customs")

    publish_ready_for_pickup_bill(
        package,
        estimated_freight_jmd=estimated_freight_jmd,
        duties_jmd=duties_jmd,
        handling_jmd=handling_jmd,
        other_fees_jmd=other_fees_jmd,
    )
    package.status = "ready_for_pickup"
    package.updated_at = datetime.utcnow()
    add_package_event(
        package,
        "ready_for_pickup",
        note or "Released from customs — bill published",
    )
    return package


def release_packages_from_customs(
    items: list[dict],
    *,
    note: str | None = None,
) -> tuple[list[Package], list[dict]]:
    released: list[Package] = []
    failed: list[dict] = []

    for item in items:
        raw_id = item.get("package_id")
        package = db.session.get(Package, raw_id)
        if not package:
            failed.append({"id": str(raw_id), "error": "Package not found"})
            continue
        try:
            release_package_from_customs(
                package,
                estimated_freight_jmd=item.get("estimated_freight_jmd"),
                duties_jmd=item.get("duties_jmd"),
                handling_jmd=item.get("handling_jmd"),
                other_fees_jmd=item.get("other_fees_jmd"),
                note=item.get("note") or note,
            )
            released.append(package)
        except ValueError as exc:
            failed.append(
                {
                    "id": str(package.id),
                    "tracking_number": package.tracking_number,
                    "error": str(exc),
                }
            )

    if released:
        db.session.commit()
    return released, failed


def bulk_request_customs_invoices(
    package_ids: list,
    channel: str,
    note: str | None = None,
) -> tuple[list[dict], list[dict]]:
    sent: list[dict] = []
    failed: list[dict] = []

    for raw_id in package_ids:
        package = db.session.get(Package, raw_id)
        if not package:
            failed.append({"id": str(raw_id), "error": "Package not found"})
            continue
        if package.status != "customs":
            failed.append(
                {
                    "id": str(package.id),
                    "tracking_number": package.tracking_number,
                    "error": "Package is not in customs",
                }
            )
            continue
        try:
            result = request_package_invoice(package, channel, note)
            sent.append(
                {
                    "package_id": str(package.id),
                    "tracking_number": package.tracking_number,
                    **result,
                }
            )
        except ValueError as exc:
            failed.append(
                {
                    "id": str(package.id),
                    "tracking_number": package.tracking_number,
                    "error": str(exc),
                }
            )

    return sent, failed

File: server/src/services/customsReleaseService.ts
import { eq, sql } from 'drizzle-orm'
import { packages } from '../db/schema/index.js'

type PackageRow = typeof packages.$inferSelect
import { db } from '../db/index.js'
import { publishReadyForPickupBill } from './billingCalculations.js'
import { requestPackageInvoice } from './billingService.js'
import { addPackageEvent } from './packageService.js'

export async function releasePackageFromCustoms(
  pkg: PackageRow,
  opts: {
    estimatedFreightJmd?: number | null
    dutiesJmd?: number | null
    handlingJmd?: number | null
    otherFeesJmd?: number | null
    note?: string | null
  } = {},
): Promise<PackageRow> {
  if (pkg.status !== 'customs') {
    throw new Error(`${pkg.trackingNumber} is not in customs`)
  }

  const mutable: PackageRow = { ...pkg }
  publishReadyForPickupBill(mutable, {
    estimatedFreightJmd: opts.estimatedFreightJmd,
    dutiesJmd: opts.dutiesJmd,
    handlingJmd: opts.handlingJmd,
    otherFeesJmd: opts.otherFeesJmd,
  })

  await db
    .update(packages)
    .set({
      estimatedFreightJmd: mutable.estimatedFreightJmd,
      dutiesJmd: mutable.dutiesJmd,
      handlingJmd: mutable.handlingJmd,
      otherFeesJmd: mutable.otherFeesJmd,
      totalDueJmd: mutable.totalDueJmd,
      billingStatus: mutable.billingStatus,
      rateTierLabel: mutable.rateTierLabel,
      billableWeightLbs: mutable.billableWeightLbs,
      updatedAt: sql`now()`,
    })
    .where(eq(packages.id, pkg.id))

  await addPackageEvent(
    mutable,
    'ready_for_pickup',
    opts.note ?? 'Released from customs — bill published',
  )

  const [updated] = await db.select().from(packages).where(eq(packages.id, pkg.id)).limit(1)
  return updated!
}

export async function releasePackagesFromCustoms(
  items: Array<Record<string, unknown>>,
  opts: { note?: string | null } = {},
): Promise<[PackageRow[], Array<Record<string, unknown>>]> {
  const released: PackageRow[] = []
  const failed: Array<Record<string, unknown>> = []

  for (const item of items) {
    const rawId = item.package_id
    const [pkg] = await db
      .select()
      .from(packages)
      .where(eq(packages.id, String(rawId)))
      .limit(1)
    if (!pkg) {
      failed.push({ id: String(rawId), error: 'Package not found' })
      continue
    }
    try {
      const updated = await releasePackageFromCustoms(pkg, {
        estimatedFreightJmd: item.estimated_freight_jmd as number | null | undefined,
        dutiesJmd: item.duties_jmd as number | null | undefined,
        handlingJmd: item.handling_jmd as number | null | undefined,
        otherFeesJmd: item.other_fees_jmd as number | null | undefined,
        note: (item.note as string | null | undefined) ?? opts.note,
      })
      released.push(updated)
    } catch (err) {
      failed.push({
        id: pkg.id,
        tracking_number: pkg.trackingNumber,
        error: err instanceof Error ? err.message : String(err),
      })
    }
  }

  return [released, failed]
}

export async function bulkRequestCustomsInvoices(
  packageIds: unknown[],
  channel: string,
  note?: string | null,
): Promise<[Array<Record<string, unknown>>, Array<Record<string, unknown>>]> {
  const sent: Array<Record<string, unknown>> = []
  const failed: Array<Record<string, unknown>> = []

  for (const rawId of packageIds) {
    const [pkg] = await db
      .select()
      .from(packages)
      .where(eq(packages.id, String(rawId)))
      .limit(1)
    if (!pkg) {
      failed.push({ id: String(rawId), error: 'Package not found' })
      continue
    }
    if (pkg.status !== 'customs') {
      failed.push({
        id: pkg.id,
        tracking_number: pkg.trackingNumber,
        error: 'Package is not in customs',
      })
      continue
    }
    try {
      const result = await requestPackageInvoice(pkg, channel, note)
      sent.push({
        package_id: pkg.id,
        tracking_number: pkg.trackingNumber,
        ...result,
      })
    } catch (err) {
      failed.push({
        id: pkg.id,
        tracking_number: pkg.trackingNumber,
        error: err instanceof Error ? err.message : String(err),
      })
    }
  }

  return [sent, failed]
}
